### ISSUE TITLE
[PSyData] Create driver for LFRic kernel extraction.

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#1483]
Create a driver for LFRic kernel extraction. This is #647 for LFRic.

**Comment by hiker** [comment]
I am currently stuck at the following issue:
the input/output list of variable signature is based on my `reference_accesses` code, and this adds all implicit variables. E.g. for a kernel taking `field1` and `field2`, I get all the variables from:
```
CALL testkern_w0_code(nlayers, field1_proxy%data, field2_proxy%data, ndf_w0, undf_w0, map_w0(:,cell))
```
E.g. cell, nlayers etc. But when my code searches for all references in the schedule, all I get is `field_proxy1` and `field_proxy2`, which I then add to the new symbol table. Later, when the driver-creator tries to create the code to read in the variable, it tries to create code for (say) `cell`, which is does not have in its symbol table.

So ... in short, atm I am stuck at the issue that the schedule after `lower_to_Language` still only lists (I think) the parameters from the meta data (field_proxy1, ...), but not the (additional) variables actually supplied (nor the fact that not field1_proxy, but field1_proxy%data is used).

Now admittedly, that is partly my fault - I added a `psyir_expressions` function to the LFRic `DynKernelArguments` class, which only returns what `self.args` contains, which I assume is just the meta-data - the two fields. I assume we can add the implicit arguments here (though I am not sure if this breaks anything else).

**Comment by hiker** [comment]
Confirmed, when I print the psyir after lower_to_language, I get:
```
subroutine invoke_testkern_w0(field1, field2)
  use testkern_w0_kernel_mod, only : testkern_w0_code
  type(field_type), intent(inout) :: field1
  type(field_type), intent(inout) :: field2
  integer :: cell

  do cell = NOT_INITIALISED, NOT_INITIALISED, 1
    call testkern_w0_code(field1_proxy, field2_proxy)
  enddo

end subroutine invoke_testkern_w0
```
But in gocean at the same stage I get:
```
subroutine invoke_1_update_field(a_fld, b_fld, c_fld, d_fld, x, y, z)
  use update_field_mod, only : update_field_code
  type(r2d_field), intent(inout) :: a_fld
  type(r2d_field), intent(inout) :: b_fld
  type(r2d_field), intent(inout) :: c_fld
  type(r2d_field), intent(inout) :: d_fld
  real(kind=go_wp), intent(inout) :: x
  real(kind=go_wp), intent(inout) :: y
  real(kind=go_wp), intent(inout) :: z
  integer :: j
  integer :: i

  do j = a_fld%whole%ystart, a_fld%whole%ystop, 1
    do i = a_fld%whole%xstart, a_fld%whole%xstop, 1
      call update_field_code(i, j, a_fld%data, b_fld%data, c_fld%data, d_fld%data, x, y, z, b_fld%grid%dx)
    enddo
  enddo

end subroutine invoke_1_update_field
```

I guess the first required step is to make `lower_to_language_level` work as expected.

**Comment by arporter** [comment]
A lot of the additional complexity in LFRic is handled by the various instances of DynCollection. This class has a method to create declarations and a method to create initialisations, both of which can probably be extended to populate the symbol table and generate Assignments, respectively.

**Comment by hiker** [comment]
In https://github.com/stfc/PSyclone/commit/21b3a25f62793f1bda7b16a93c618c1b68375d39 I've reached the stage to get:
```
subroutine invoke_testkern_w0(field1, field2)
  use testkern_w0_kernel_mod, only : testkern_w0_code
  type(field_type), intent(inout) :: field1
  type(field_type), intent(inout) :: field2
  integer :: cell

  do cell = NOT_INITIALISED, NOT_INITIALISED, 1
    call testkern_w0_code(field1_proxy%data, field2_proxy%data)
  enddo

end subroutine invoke_testkern_w0
```
Not a big difference, but at least we get the correct structure reference now (though field1_proxy etc are still not declared). I started to look at KernCallArgList, and we could try to return a list of References (instead of string), which we could then use in `DynKernelArguments.psyir_expressions`. But the devil is in the detail:
1. KernCallARgList is also used for variable access determination. We then need to map the references to Signatures (which is reasonable easy to do, all StructureReferences already have a method for that). But
2. Who is responsible for actually declaring the symbols? Imho the symbols should be declared in the code that uses them (e.g. that assigned to a field_proxy variable). I am not convinced that the type information should be in the kernel argument classes (they should rely on the symbol already being in the symbol table I think).
